tensorflow/core/common_runtime/gpu/gpu_cudamallocasync_allocator.cc

```diff
@@ -0,0 +1,207 @@
+  // case for TF+horovod.
+  if (platform_gpu_id.value() > 0) {
+    CUcontext pctx; // We loose track of it. But this is fine.
+    CUresult err = cuDevicePrimaryCtxRetain(&pctx, 0);
+    if (err != CUDA_SUCCESS){
+      LOG(FATAL) << "Failed to create the context on device 0.";
+    }
+  }
+
+  se::cuda::ScopedActivateExecutorContext scoped_activation{stream_exec_};
+  int cuda_malloc_async_supported;
+  cudaDeviceGetAttribute(&cuda_malloc_async_supported,
+                         cudaDevAttrMemoryPoolsSupported,
+                         platform_gpu_id.value());
+  if (!cuda_malloc_async_supported) {
+    LOG(FATAL) << "TF_GPU_ALLOCATOR=cuda_malloc_async isn't currently supported."
+               << " Possible causes: device not supported, driver too old, "
+               << " OS not supported, CUDA version too old.";
+  }
+
+  cudaError_t cerr = cudaStreamCreate(&cuda_stream_);
+  if (cerr != cudaSuccess) {
+    LOG(FATAL) << "could not allocate CUDA stream for context : "
+               << cudaGetErrorString(cerr);
+  }
+
+  cerr = cudaDeviceGetDefaultMemPool(&pool_, platform_gpu_id.value());
+  if (cerr != cudaSuccess) {
+    LOG(FATAL) << "could not get the default CUDA pool : "
+               << cudaGetErrorString(cerr);
+  }
+  VLOG(1) << Name() << " CudaMallocAsync initialized on platform: "
+          << platform_gpu_id.value() << " with pool size of: "
+          << pool_size << " this ptr: " << this;
+  cerr = cudaMemPoolSetAttribute(pool_, cudaMemPoolAttrReleaseThreshold,
+                                 (void*)&pool_size);
+  if (compute_stats) {
+    stats_.bytes_limit = static_cast<int64>(pool_size);
+  } // If not set, it means we do not compute stats.
+  if (cerr != cudaSuccess) {
+    LOG(FATAL) << "could not set the default CUDA pool memory threshold : "
+               << cudaGetErrorString(cerr);
+    pool_ = nullptr;
+  }
+
+  // If in TF_DETERMINISTIC_OPS is set, then make the allocator behave
+  // determistically.
+  bool deterministic_ops = false;
+  TF_CHECK_OK(tensorflow::ReadBoolFromEnvVar("TF_DETERMINISTIC_OPS",
+                                             /*default_val=*/false,
+                                             &deterministic_ops));
+  if (deterministic_ops) {
+    cudaMemPoolSetAttribute(pool_, cudaMemPoolReuseAllowOpportunistic, 0);
+    cudaMemPoolSetAttribute(pool_, cudaMemPoolReuseAllowInternalDependencies, 0);
+  }
+#endif
+
+  VLOG(2) << Name() << " GPUcudaMallocAsyncAllocator PoolSize " << pool_size;
+  if (reserve_memory) {
+    void* ptr = AllocateRaw(0, pool_size);
+    DeallocateRaw(ptr);
+    VLOG(2) << Name() << " GPUcudaMallocAsyncAllocator reserved the pool";
+    ClearStats();
+  }
+}
+
+GPUcudaMallocAsyncAllocator::~GPUcudaMallocAsyncAllocator() {
+  cuStreamDestroy(cuda_stream_);
+}
+
+void* GPUcudaMallocAsyncAllocator::AllocateRaw(size_t alignment,
+                                               size_t num_bytes) {
+#if CUDA_VERSION < 11020 || !defined(GOOGLE_CUDA)
```

Comment:
Can you move the ifdef guards outside the class so that you don't need the bogus pointers? Perhaps add a warning in useCudaMallocAsyncAllocator that informs the user that env variable has no effect because they're using the wrong version of CUDA